core/src/main/java/ai/timefold/solver/core/impl/bavet/common/AbstractPrecomputeNode.java

```diff
@@ -0,0 +1,184 @@
+package ai.timefold.solver.core.impl.bavet.common;
+
+import java.util.ArrayList;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import ai.timefold.solver.core.impl.bavet.NodeNetwork;
+import ai.timefold.solver.core.impl.bavet.common.tuple.AbstractTuple;
+import ai.timefold.solver.core.impl.bavet.common.tuple.RecordingTupleLifecycle;
+import ai.timefold.solver.core.impl.bavet.common.tuple.TupleLifecycle;
+import ai.timefold.solver.core.impl.bavet.common.tuple.TupleState;
+import ai.timefold.solver.core.impl.bavet.uni.AbstractForEachUniNode;
+import ai.timefold.solver.core.impl.util.CollectionUtils;
+
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+public abstract class AbstractPrecomputeNode<Tuple_ extends AbstractTuple> extends AbstractNode
+        implements BavetRootNode<Object> {
+    private final StaticPropagationQueue<Tuple_> propagationQueue;
+    private final Map<Object, List<Tuple_>> objectToOutputTuplesMap = new IdentityHashMap<>(1000);
+    private final IdentityHashMap<Tuple_, Tuple_> inputTupleToOutputTupleMap = new IdentityHashMap<>(1000);
+    private final NodeNetwork innerNodeNetwork;
+    private final RecordingTupleLifecycle<Tuple_> recordingTupleNode;
+    private final Class<?>[] sourceClasses;
+    private final Set<Object> queuedInsertSet = CollectionUtils.newIdentityHashSet(32);
+    private final Set<Object> queuedUpdateSet = CollectionUtils.newIdentityHashSet(32);
+    private final Set<Object> queuedRetractSet = CollectionUtils.newIdentityHashSet(32);
+
+    protected AbstractPrecomputeNode(NodeNetwork innerNodeNetwork,
+            RecordingTupleLifecycle<Tuple_> recordingTupleNode,
+            TupleLifecycle<Tuple_> nextNodesTupleLifecycle,
+            Class<?>[] sourceClasses) {
+        this.innerNodeNetwork = innerNodeNetwork;
+        this.propagationQueue = new StaticPropagationQueue<>(nextNodesTupleLifecycle);
+        this.sourceClasses = sourceClasses;
+        this.recordingTupleNode = recordingTupleNode;
+    }
+
+    @Override
+    public final Propagator getPropagator() {
+        return propagationQueue;
+    }
+
+    @Override
+    public final boolean allowsInstancesOf(Class<?> clazz) {
+        for (var sourceClass : sourceClasses) {
+            if (sourceClass.isAssignableFrom(clazz)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public final Class<?>[] getSourceClasses() {
+        return sourceClasses;
+    }
+
+    @Override
+    public final boolean supports(BavetRootNode.LifecycleOperation lifecycleOperation) {
+        return true;
+    }
+
+    @Override
+    public final void insert(Object a) {
+        // do not remove a retract of the same fact (a fact was updated)
+        queuedInsertSet.add(a);
+    }
+
+    @Override
+    public final void update(Object a) {
+        queuedUpdateSet.add(a);
```

Comment:
I wonder... should we prevent the update, if we see that `a` is already inserting? In that case, we only need to track the insert.